### PR TITLE
drivers: digital-io: max22190 : Add max22190 driver support

### PR DESCRIPTION
**Pull request by RaduSabau1**
## Pull Request Description

MAX22190 is an octal industrial digital input device with diagnostics that communicates through a SPI interface.
The files added by this PR have been tested with the MAX22190 PMB connected to a max32690 microcontroller.
The driver has the basic api for init, remove, register read, write and update, but also a custom CRC5 algorithm specific for this device 
alongside its specific api(filter_get, filter_set, chan_state).

## PR Type
- [ ] Bug fix (change that fixes an issue)
- [x] New feature (change that adds new functionality)
- [ ] Breaking change (has dependencies in other repos or will cause CI to fail)

## PR Checklist
- [x] I have followed the [Coding style guidelines](http://analogdevicesinc.github.io/no-OS/drivers_guide.html#coding-style)
- [ ] I have performed a self-review of the changes
- [x] I have commented my code, at least hard-to-understand parts
- [ ] I have build all projects affected by the changes in this PR
- [ ] I have tested in hardware affected projects, at the relevant boards
- [x] I have signed off all commits from this PR
- [ ] I have updated the documentation (wiki pages, ReadMe etc), if applies

### CODE REVIEW

**Review by buha — APPROVED** [review]

File: drivers/digital-io/max22190/iio_max22190.h
Comment (buha) [inline drivers/digital-io/max22190/iio_max22190.h:49]: please document in a comment what this acronyms stand for: `wbe` and `fbp`

File: drivers/digital-io/max22190/iio_max22190.c
Comment (buha) [inline drivers/digital-io/max22190/iio_max22190.c:141]: as discussed, this approach is a bit ugly but iio doesn't provide a proper way for exposing the status that comes with the data, so from my point of view it's good, maybe team has a better idea ?

File: drivers/digital-io/max22190/iio_max22190.c
Comment (buha) [inline drivers/digital-io/max22190/iio_max22190.c:386]: idk about these blank lines... i think we'd be better without

File: drivers/digital-io/max22190/iio_max22190.h
Comment (CiprianRegus) [inline drivers/digital-io/max22190/iio_max22190.h:60]: Unused

File: drivers/digital-io/max22190/iio_max22190.h
Comment (CiprianRegus) [inline drivers/digital-io/max22190/iio_max22190.h:62]: Why do you need to store this?

File: drivers/digital-io/max22190/iio_max22190.h
Comment (CiprianRegus) [inline drivers/digital-io/max22190/iio_max22190.h:70]: You're not using this.

File: drivers/digital-io/max22190/iio_max22190.c
Comment (CiprianRegus) [inline drivers/digital-io/max22190/iio_max22190.c:179]: This is not for a specific channel, you're just storing the value of a register.

File: drivers/digital-io/max22190/max22190.c
Comment (CiprianRegus) [inline drivers/digital-io/max22190/max22190.c:101]: off by 1. Check throughout the file.

File: drivers/digital-io/max22190/max22190.c
Comment (CiprianRegus) [inline drivers/digital-io/max22190/max22190.c:199]: What is `0x1F`

File: drivers/digital-io/max22190/max22190.h
Comment (CiprianRegus) [inline drivers/digital-io/max22190/max22190.h:87]: You're not using it.

File: drivers/digital-io/max22190/max22190.c
Comment (CiprianRegus) [inline drivers/digital-io/max22190/max22190.c:295]: What does this do? Adding a macro for the mask may help.

File: drivers/digital-io/max22190/max22190.h
Comment (RaduSabau1) [reply drivers/digital-io/max22190/max22190.h:87]: I am using it in the max22190_chan_state function.

File: drivers/digital-io/max22190/max22190.c
Comment (RaduSabau1) [reply drivers/digital-io/max22190/max22190.c:295]: A reg_update isn't really needed on that specific mask, so I will replace it with a simple register read to clear the latched faults.

File: drivers/digital-io/max22190/max22190.c
Comment (RaduSabau1) [reply drivers/digital-io/max22190/max22190.c:101]: will fix it.

File: drivers/digital-io/max22190/iio_max22190.h
Comment (RaduSabau1) [reply drivers/digital-io/max22190/iio_max22190.h:49]: Will document them

File: drivers/digital-io/max22190/iio_max22190.h
Comment (RaduSabau1) [reply drivers/digital-io/max22190/iio_max22190.h:60]: Will remove

File: drivers/digital-io/max22190/iio_max22190.h
Comment (RaduSabau1) [reply drivers/digital-io/max22190/iio_max22190.h:62]: I was using it in an application example, but by taking a closer look, i realize this should be project specific instead of it being implemented in the driver itself.

File: drivers/digital-io/max22190/iio_max22190.c
Comment (RaduSabau1) [reply drivers/digital-io/max22190/iio_max22190.c:179]: Will use no_os_field_get, my bad.

File: drivers/digital-io/max22190/iio_max22190.c
Comment (RaduSabau1) [reply drivers/digital-io/max22190/iio_max22190.c:386]: Will remove them.

File: drivers/digital-io/max22190/iio_max22190.h
Comment (RaduSabau1) [reply drivers/digital-io/max22190/iio_max22190.h:49]: Actually, struct max22190_ch_filter will be removed, because of it being used only in the specific application example i created to test this on.
In case of a default mode filters are bypassed, therefore this struct wouldn't make sense.

File: drivers/digital-io/max22190/iio_max22190.c
Comment (RaduSabau1) [reply drivers/digital-io/max22190/iio_max22190.c:141]: Will try this instead.

File: drivers/digital-io/max22190/max22190.h
Comment (CiprianRegus) [reply drivers/digital-io/max22190/max22190.h:87]: It's better to not duplicate your device's state in the driver. Just read the status register each time.

**Comment by RaduSabau1** [comment]
V2

- Removed unused variables.
- Modified driver init function.
- fault1 and fault2 are debug attributes now.
- Fixed read_raw value to be masked for the specific channel.
- Addressed other comments.

**Comment by RaduSabau1** [comment]
V3

- POR bit is now cleared on initialization(fault1 register is read because of the reg_update api call in the init).

File: drivers/digital-io/max22190/max22190.h
Comment (buha) [inline drivers/digital-io/max22190/max22190.h:54]: paranthesize x as `(x)`

Comment (buha) [reply drivers/digital-io/max22190/max22190.h:54]: missing whitespaces

**Comment by RaduSabau1** [comment]
V4

- Fixed MACRO definition bug.

**Review by amiclaus — APPROVED** [review]

File: drivers/digital-io/max22190/README.rst
Comment (buha) [inline drivers/digital-io/max22190/README.rst:70]: please put hardcoded values here, proper examples, like you did for 14916

**Comment by RaduSabau1** [comment]
V5

- Updated README.

File: drivers/digital-io/max22190/README.rst
Comment (buha) [inline drivers/digital-io/max22190/README.rst:95]: This doesn't render correctly, please test sphinx output

Also, you don't need to backtick quote. Perhaps use it on the attribute name to render like this (but in sphinx of course):
* `raw` - raw value of a channel

Comment (RaduSabau1) [reply drivers/digital-io/max22190/README.rst:95]: will test the sphinx output, and make sure it renders correctly.

**Comment by RaduSabau1** [comment]
V6

- Updated README.

**Comment by RaduSabau1** [comment]
V7

- Updated IIO driver(switched to channel's address instead of ch_num).

File: drivers/digital-io/max22190/README.rst
Comment (rbolboac) [inline drivers/digital-io/max22190/README.rst:1]: you should also source this file in sphinx/source similarly to what it is already done for other drivers

**Comment by RaduSabau1** [comment]
V8

- Included README.rst in sphinx/source.

File: doc/sphinx/source/drivers_doc.rst
Comment (amiclaus) [inline doc/sphinx/source/drivers_doc.rst:24]: after the merge of #1955, you need only `drivers/max22190` to add, since the ToC is already created. (therefore the newly appeared conflict.

Comment (RaduSabau1) [reply doc/sphinx/source/drivers_doc.rst:24]: I thought about this when writing the code. Thinking of waiting for #1955 to be merged first, and then only add drivers/max22190 line in this file (since ToC will already be created).

Comment (amiclaus) [reply doc/sphinx/source/drivers_doc.rst:24]: I merged it so you can resolve already the conflict :)

Comment (RaduSabau1) [reply doc/sphinx/source/drivers_doc.rst:24]: On it.

**Review by buha — APPROVED** [review]

File: drivers/digital-io/max22190/iio_max22190.h
Comment (rbolboac) [inline drivers/digital-io/max22190/iio_max22190.h:71]: missing new line at the end of file, other than that can be merged

**Comment by RaduSabau1** [comment]
V10

- Added new line at the EOF (iio_max22190.h).

**Review by amiclaus — APPROVED** [review]

**Review by rbolboac — APPROVED** [review]

File: drivers/digital-io/max22190/README.rst
Comment (CiprianRegus) [inline drivers/digital-io/max22190/README.rst:94]: These are channel attributes

File: drivers/digital-io/max22190/iio_max22190.c
Comment (CiprianRegus) [inline drivers/digital-io/max22190/iio_max22190.c:150]: Use just 1 function and set different `priv` values for each of these struct elements.

File: drivers/digital-io/max22190/iio_max22190.c
Comment (CiprianRegus) [inline drivers/digital-io/max22190/iio_max22190.c:481]: Is there any reason why wouldn't all channels be enabled?

Comment (RaduSabau1) [reply drivers/digital-io/max22190/iio_max22190.c:481]: No, will remove the boolean array and everything related to it.

File: drivers/digital-io/max22190/iio_max22190.c
Comment (RaduSabau1) [reply drivers/digital-io/max22190/iio_max22190.c:150]: Will do.

**Comment by RaduSabau1** [comment]
V12

- Fixed README.

**Comment by RaduSabau1** [comment]
V13

- iio_channels are now declared static and iio_setup_channels is removed.

**Comment by RaduSabau1** [comment]
V14

- Fixed error for iio_driver in the iio_device structure.
- Fixed warnings

File: drivers/digital-io/max22190/max22190.c
Comment (rbolboac) [inline drivers/digital-io/max22190/max22190.c:179]: I think this switch case can be rewritten with less lines in a more generic way.
something like:
```
case MAX22190_FAULT1_WBGE ... MAX22190_FAULT1_FAULT2E:
ret = max22190_reg_read(desc, MAX22190_FAULT1_EN_REG, &reg_val);
		if (ret)
			return ret;
		*enabled = no_os_field_get(MAX22190_FAULT_MASK(fault_enable), reg_val);
		break;
```
where
```
#define MAX22190_FAULT_MASK(x) NO_OS_BIT(x)
```

similarly for fault 2, you will only have to take care of the enum offset:
```
case MAX22190_FAULT2_RFWBSE ... MAX22190_FAULT2_FAULT8CKE:
                ret = max22190_reg_read(desc, MAX22190_FAULT2_EN_REG, &reg_val);
		if (ret)
			return ret;
		*enabled = no_os_field_get(MAX22190_FAULT_MASK(fault_enable-MAX22190_FAULT2_RFWBSE), reg_val);
		break;
```

File: drivers/digital-io/max22190/max22190.c
Comment (rbolboac) [inline drivers/digital-io/max22190/max22190.c:172]: let's add similar APIs for other registers (e.g. config register, filters, etc)

**Comment by RaduSabau1** [comment]
V15

- Added API for fault_enable and wbe.
- Added debug attributes for fault_enable in the IIO driver.
- Modified README.rst.

File: drivers/digital-io/max22190/max22190.c
Comment (RaduSabau1) [reply drivers/digital-io/max22190/max22190.c:231]: Config reg has 2 bits that can be added to fault and 1 for register, and I will do exactly that.

**Comment by RaduSabau1** [comment]
V16

- Modified documentation to match the driver.

File: drivers/digital-io/max22190/max22190.c
Comment (rbolboac) [inline drivers/digital-io/max22190/max22190.c:221]: remove extra semicolon

File: drivers/digital-io/max22190/max22190.c
Comment (rbolboac) [inline drivers/digital-io/max22190/max22190.c:291]: can use `...` syntax similarly to get function

File: drivers/digital-io/max22190/iio_max22190.c
Comment (rbolboac) [inline drivers/digital-io/max22190/iio_max22190.c:294]: if the offset is always 0, why have it at all?

File: drivers/digital-io/max22190/iio_max22190.c
Comment (rbolboac) [inline drivers/digital-io/max22190/iio_max22190.c:431]: code can be optimized by taking this if condition outside of for loop

**Comment by RaduSabau1** [comment]
V17

- Optimized code.
- Removed offset.
- Addressed other comments.

File: drivers/digital-io/max22190/max22190.c
Comment (buha) [inline drivers/digital-io/max22190/max22190.c:291]: so what if I want to enabled MAX22190_FAULT2_RFWBSE ?
`MAX22190_FAULT1_FAULT2E` should probably be automatically enabled by the driver, it is currently not

Comment (buha) [reply drivers/digital-io/max22190/max22190.c:291]: This would imply keeping track of fault2 enables somewhere in desc->fault2en or something like that.
When any fault2 enable gets enabled, `MAX22190_FAULT1_FAULT2E` should be set.
When no fault2 enable remains, `MAX22190_FAULT1_FAULT2E` should be cleared.

Comment (RaduSabau1) [reply drivers/digital-io/max22190/max22190.c:292]: I'll do that but having a fault2en in the desc isn't necessary.

**Comment by RaduSabau1** [comment]
V18

- Checking fault2 enable in fault1 register when setting fault 2 enable.

**Comment by RaduSabau1** [comment]
V19

- Added fault2en in device descriptor and modified fault_enable set API in order to optimize it.

File: drivers/digital-io/max22190/max22190.c
Comment (buha) [inline drivers/digital-io/max22190/max22190.c:304]: probably works as-is but if fault2en would be a bitmask, you would not need this for loop

**Comment by RaduSabau1** [comment]
V20

- Modified fault2en to be a bit mask.

File: drivers/digital-io/max22190/max22190.c
Comment (RaduSabau1) [reply drivers/digital-io/max22190/max22190.c:514]: it's not, it will be removed.

**Comment by RaduSabau1** [comment]
V21

- Removed unnecessary lines.

**Comment by RaduSabau1** [comment]
V21

- Removed unnecessary lines.

**Review by rbolboac — APPROVED** [review]

**Pull request merged**
